File: Day029_2025/Q029_GFG.cpp
//{ Driver Code Starts
// Initial Template for C++
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++
class Solution {
  public:
    double power(double b, int e) {
        // Q. Implement Pow
        // code here
        if (e == 0){
            return 1;
        }
        
        if (e < 0){
            return 1 / power(b, -e);
        }
            
        double halfPower = power(b, e / 2);
    	
        if (e % 2 == 0){
            return halfPower * halfPower;
        }
            
        else{
            return b * halfPower * halfPower;
        }
    }
};

//{ Driver Code Starts.

int main() {
    cout << fixed << setprecision(5);
    int t;
    cin >> t;
    while (t--) {
        double b;
        cin >> b;
        int e;
        cin >> e;
        Solution ob;
        cout << ob.power(b, e) << endl;
        cout << "~" << endl;
    }

    return 0;
}
// } Driver Code Ends

File: Day029_2025/Q029_GFG.py
class Solution:
    def power(self, b: float, e: int) -> float:
        if e == 0:
            return 1.0
        
        if e < 0:
            b = 1 / b
            e = -e

        halfPower = self.power(b, e // 2)

        if e % 2 == 0:
            return halfPower * halfPower
        else:
            return b * halfPower * halfPower
